File: SDI/Chat RabbitMq/Cliente.py
#!/usr/bin/env python
import pika
import sys
from threading import Thread
import os

nome = input("Digite o seu nome:")
cont_recebido = 0

class T(Thread):

	def __init__ (self, channel_saida_, nome_):
		Thread.__init__(self)
		self.channel_saida = channel_saida_
		self.nome = nome_
		self.cont_enviado = 0
		self.cont_recebido = 0

	def run(self):cd 
		envia_saida(self.channel_saida, self.nome, self.cont_enviado)
 
def main():

	cont_enviado = 0

	print("\nPara enviar uma mensagem digite o nome do arquivo:\n")

	#connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
	connection = pika.BlockingConnection(pika.ConnectionParameters(
        host='sdi-1',
        credentials=pika.PlainCredentials(username="sdi", password="sdi")))
	channel_saida = connection.channel()
	channel_saida.exchange_declare(exchange='saida', exchange_type='fanout')

	channel_entrada = connection.channel()
	channel_entrada.exchange_declare(exchange='entrada', exchange_type='fanout')
	result = channel_entrada.queue_declare('', exclusive=True)
	queue_name = result.method.queue
	channel_entrada.queue_bind(exchange='entrada', queue=queue_name)

	channel_entrada.basic_consume(callback, queue=queue_name, no_ack=True)

	t = T(channel_saida, nome)
	t.start()

	channel_entrada.start_consuming()


def callback(ch, method, properties, body):
	message = body.decode().split("&", 2)
	if(message[0]!=nome):
		n_arq = nome+"-"+message[0]+"-"+message[1]+".client"
		arq = open(n_arq, 'w')
		arq.write(message[2])
		print("Recebi uma mensagem de " +message[0]+". Salvei no arquivo: %r" % n_arq)
		print("\nPara enviar uma mensagem digite o nome do arquivo:")

def envia_saida(channel_saida, nome, cont_enviado):
	
	while(True):
		n_arquivo = input("")
		if os.path.isfile(n_arquivo):
			arq = open(n_arquivo, 'r')
			message = arq.read()
			message = nome+"&"+str(cont_enviado)+"&"+message
			cont_enviado+=1
			channel_saida.basic_publish(exchange='saida', routing_key='', body=message)
			print("Enviei a mensagem do arquivo "+n_arquivo+".\n\nPara enviar uma mensagem digite o nome do arquivo:")
		else:
			print("Nao existe arquivo com esse nome.\n\nPara enviar uma mensagem digite o nome do arquivo:")

#connection.close()

if __name__ == "__main__":
    main()

File: SDI/Chat RabbitMq/Cliente.java
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
//import com.rabbitmq.client.DeliverCallback;
import com.rabbitmq.client.*;
import java.util.*;
import java.io.*;
import java.nio.file.*;


public class Cliente implements Runnable{

	static String nome;

	public Cliente(){}
	public Cliente(String nome_){
		this.nome = nome_;
	}
	
	static int cont_recebido = 0, cont_enviado = 0;
	static Scanner scan = new Scanner(System.in);

	static String EXCHANGE_ENTRADA = "entrada";
	static ConnectionFactory factory; 
    static Connection connection;
    static Channel channel_entrada;

    static String queueName;

    static String EXCHANGE_SAIDA = "saida";
    static Channel channel_saida;

    static String n_arquivo = null;

	public static void main(String args[]){

		try{
			factory = new ConnectionFactory();
	    	factory.setHost("localhost");
	    	connection = factory.newConnection();
	    	channel_entrada = connection.createChannel();
	    	channel_entrada.exchangeDeclare(EXCHANGE_ENTRADA, "fanout");
	    	queueName = channel_entrada.queueDeclare().getQueue();
	    	channel_entrada.queueBind(queueName, EXCHANGE_ENTRADA, "");
		
	    	channel_saida = connection.createChannel();
    		channel_saida.exchangeDeclare(EXCHANGE_SAIDA, "fanout");
		} catch(Exception ee){}

		System.out.println("Digite o seu nome:");
		nome = scan.nextLine();
		System.out.println("\nPara enviar uma mensagem digite o nome do arquivo:\n");
		new Thread(new Cliente(nome)).start();
		File file;

		//DeliverCallback deliverCallback = (consumerTag, delivery) -> {
        Consumer consumer = new DefaultConsumer(channel_entrada){
        public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body){
        	//String message = new String(delivery.getBody(), "UTF-8");
        	String message = new String(body);
        	String aux[] = message.split("&", 3);
        	if(!aux[0].equals(nome)){
        		n_arquivo = nome+"-"+aux[0]+"-"+aux[1]+".client";
        		Cliente.write(n_arquivo, aux[2]);
        		System.out.println("Recebi uma mensagem de " +aux[0]+". Salvei no arquivo: "+n_arquivo);
				System.out.println("\nPara enviar uma mensagem digite o nome do arquivo:");
        	}
        }
    	};
		
		try{
			channel_entrada.basicConsume(queueName, true, consumer);
		} catch (Exception e){}
		

	}

	public void run(){

		String n_arq;
		File file;
		String message;
		while(true){
			n_arq = scan.nextLine();
			file = new File(n_arq);
			if(file.exists()){
				message = Cliente.read(n_arq);
				message = this.nome+"&"+cont_enviado+"&"+message;
				cont_enviado++;
				try{
					channel_saida.basicPublish(EXCHANGE_SAIDA, "", null, message.getBytes("UTF-8"));
				} catch (Exception e){}
				System.out.println("Enviei a mensagem do arquivo "+n_arq+".\n\nPara enviar uma mensagem digite o nome do arquivo:");
			}else{
				System.out.println("Nao existe arquivo com esse nome.\n\nPara enviar uma mensagem digite o nome do arquivo:");
			}
		}

	}

	public static void write(String n_arq, String mensagem) {
	    BufferedWriter bw = null;
	    FileWriter fw = null;
	    try {
	      fw = new FileWriter(n_arq);
	      bw = new BufferedWriter(fw);
	      bw.write(mensagem);
	    } catch (IOException e) {
	      e.printStackTrace();
	    } finally {
	      try {
	        if (bw != null)
	          bw.close();
	        if (fw != null)
	          fw.close();
	      } catch (IOException ex) {
	        ex.printStackTrace();
	      }
	    }
	}

	public static String read(String arquivo) {
        BufferedReader br = null;
        FileReader fr = null;
        String FILENAME = arquivo;
        String resposta = "";
        int i=0;
        try {
            fr = new FileReader(FILENAME);
            br = new BufferedReader(fr);
            String sCurrentLine;
            br = new BufferedReader(new FileReader(FILENAME));
            while ((sCurrentLine = br.readLine()) != null) {
                if(i==0){
                    resposta = resposta+sCurrentLine;
                    i++;
                }
                else
                    resposta = resposta+"\n"+sCurrentLine;
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (br != null) {
                   br.close();
                }
                if (fr != null) {
                    fr.close();
                }
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
        return resposta;
    }
}